[14.0][FIX] l10n br fiscal: ncm com inicio 09 conflita com tipo servico

crsilveira opened the pull request:
Issue: #3095

crsilveira committed:
Merge pull request #1 from OCA/14.0

atualizando 03-2024

crsilveira committed:
Merge pull request #2 from OCA/14.0

atualização

WesleyOliveira98 committed:
[FIX] l10n_br_purchase: fix purchase order line form button 'save and new'

crsilveira committed:
Merge pull request #3 from OCA/14.0

atualização maio/24

crsilveira committed:
como estava não permitia o uso do ncm com  inicio 09.

antoniospneto commented on l10n_br_fiscal/models/product_mixin.py:
```diff
@@ -32,5 +32,5 @@ def _onchange_ncm_id(self):
                 )
 
-            if r.fiscal_genre_id.code == PRODUCT_FISCAL_TYPE_SERVICE:
+            if r.fiscal_type == PRODUCT_FISCAL_TYPE_SERVICE:
```
acho que vc tem que alterar o decorator do metodo tbm, trocar o `fiscal_genre_id` por `fiscal_type`

antoniospneto commented:
@crsilveira  depois precisa arrumar os commits tbm, juntar em um só, use o opção de squash no rebase iterativo 
e renomear com o padrão oca, que nesse caso anotar com a tag "[FIX]" veja:
https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#commit-message

marcelsavegnago committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 12.3% (73 of 591 strings)

Translation: l10n-brazil-14.0/l10n-brazil-14.0-l10n_br_nfe
Translate-URL: https://translation.odoo-community.org/projects/l10n-brazil-14-0/l10n-brazil-14-0-l10n_br_nfe/pt_BR/

marcelsavegnago committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 0.9% (11 of 1134 strings)

Translation: l10n-brazil-14.0/l10n-brazil-14.0-l10n_br_nfe_spec
Translate-URL: https://translation.odoo-community.org/projects/l10n-brazil-14-0/l10n-brazil-14-0-l10n_br_nfe_spec/pt_BR/

marcelsavegnago committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 78.2% (18 of 23 strings)

Translation: l10n-brazil-14.0/l10n-brazil-14.0-l10n_br_crm
Translate-URL: https://translation.odoo-community.org/projects/l10n-brazil-14-0/l10n-brazil-14-0-l10n_br_crm/pt_BR/

marcelsavegnago committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 92.3% (277 of 300 strings)

Translation: l10n-brazil-14.0/l10n-brazil-14.0-l10n_br_purchase
Translate-URL: https://translation.odoo-community.org/projects/l10n-brazil-14-0/l10n-brazil-14-0-l10n_br_purchase/pt_BR/

marcelsavegnago committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (274 of 274 strings)

Translation: l10n-brazil-14.0/l10n-brazil-14.0-l10n_br_contract
Translate-URL: https://translation.odoo-community.org/projects/l10n-brazil-14-0/l10n-brazil-14-0-l10n_br_contract/pt_BR/

marcelsavegnago committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (298 of 298 strings)

Translation: l10n-brazil-14.0/l10n-brazil-14.0-l10n_br_repair
Translate-URL: https://translation.odoo-community.org/projects/l10n-brazil-14-0/l10n-brazil-14-0-l10n_br_repair/pt_BR/

marcelsavegnago committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 98.6% (288 of 292 strings)

Translation: l10n-brazil-14.0/l10n-brazil-14.0-l10n_br_stock_account
Translate-URL: https://translation.odoo-community.org/projects/l10n-brazil-14-0/l10n-brazil-14-0-l10n_br_stock_account/pt_BR/

marcelsavegnago committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 90.4% (1257 of 1389 strings)

Translation: l10n-brazil-14.0/l10n-brazil-14.0-l10n_br_fiscal
Translate-URL: https://translation.odoo-community.org/projects/l10n-brazil-14-0/l10n-brazil-14-0-l10n_br_fiscal/pt_BR/

marcelsavegnago committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 92.0% (300 of 326 strings)

Translation: l10n-brazil-14.0/l10n-brazil-14.0-l10n_br_sale
Translate-URL: https://translation.odoo-community.org/projects/l10n-brazil-14-0/l10n-brazil-14-0-l10n_br_sale/pt_BR/

marcelsavegnago committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 70.5% (342 of 485 strings)

Translation: l10n-brazil-14.0/l10n-brazil-14.0-l10n_br_pos
Translate-URL: https://translation.odoo-community.org/projects/l10n-brazil-14-0/l10n-brazil-14-0-l10n_br_pos/pt_BR/

marcelsavegnago committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (277 of 277 strings)

Translation: l10n-brazil-14.0/l10n-brazil-14.0-l10n_br_sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/l10n-brazil-14-0/l10n-brazil-14-0-l10n_br_sale_blanket_order/pt_BR/

marcelsavegnago commented:
Ontem passei estas instruções em uma outra PR...  segue como exemplo.

Here are the steps to do that:

1. **Start an interactive rebase:**
   In your terminal, navigate to the root directory of your repository and run:
   ```bash
   git rebase -i HEAD~3
   ```
   This command opens the last three commits for interactive rebase.

2. **Edit the rebase list:**
   Change "pick" to "squash" for the commits you want to combine into the first commit. Your editor will show something like this:
   ```plaintext
   pick e73a481 Remove field groups_id from res_partner_view.xml
   squash df28461 [MIG] helpdesk_mgmt_partner_sequence: Migration to 16.0
   squash 625e0fa Merge branch '16.0-mig-helpdesk_mgmt_partner_sequence' of https://git...
   ```

3. **Save and close the editor:**
   After changing the lines, save and close your text editor. Git will combine the commits into one.

4. **Edit the commit message:**
   Git will prompt you to edit the commit message for the combined commit. Combine the messages as needed:
   ```plaintext
   [MIG] helpdesk_mgmt_partner_sequence: Migration to 16.0
   
   - Remove field groups_id from res_partner_view.xml
   - Merge branch '16.0-mig-helpdesk_mgmt_partner_sequence'
   ```

5. **Save and close the editor again:**
   Save the new commit message and close the editor.

6. **Force push to update the remote repository:**
   Since you've rewritten the commit history, you need to force push to the remote repository:
   ```bash
   git push REMOTE_NAME BRANCH_NAME --force
   ```

By following these steps, you will squash the specified commits into a single migration commit and update the remote repository accordingly.

OCA-git-bot committed:
[BOT] post-merge updates

rvalyi commented:
Ola @crsilveira vc consegue fazer o rebase e fazer o squash por favor?

Talvez vc tem a impressão que nao faz diferença, mas para a gente que mantem o projeto te garanto que faz uma baita diferença...

Sempre que tem algum bug cabeludo e que a gente não sabe bem porque uma linha de codigo esta do jeito que esta a gente mergulha num "git blame" para analisar o histórico e cada commit inutil atrapalha...

te do outro examplo, esse é o tipo de malabarismo que temos que fazer na OCA para mandar cada commit de uma branch para a outra, por exemplo entre as branches 14.0 e 16.0 para migrar o módulo l10n_br_account:
https://github.com/OCA/l10n-brazil/pull/2865#issuecomment-2130737216

Então se puder aprender a fazer o rebase e squash ajudaria... Até para vc é fondamental vc dominar o rebase para sobreviver com Odoo...

crsilveira commented:
Fechando aqui, pra tentar fazer corretamente

crsilveira commented:
Uma dúvida:  o que é o correto pra ficar neste decorator  :  " ...decorator do metodo tbm, trocar o fiscal_genre_id por fiscal_type "   ???

antoniospneto commented:
> Uma dúvida: o que é o correto pra ficar neste decorator : " ...decorator do metodo tbm, trocar o fiscal_genre_id por fiscal_type " ???

No decorator vc coloca os campos que vão "acionar" essa função do onchange, no caso da forma que tá hoje é sempre que alterar o campo `ncm_id` ou `fiscal_genre_id`  
mas agora não faz sentido acionar pelo o `fiscal_genre_id` tem que acionar quando o valor de `fiscal_type` mudar que é o que você tá verificando no IF agora.

Olhando melhor vi que já existe um onchange exclusivo para o `fiscal_type` acho que o mais correto é você mover essa atribuição do ncm de serviço diretamente ali no onchange do `fiscal_type`. depois apagar tbm o `fiscal_genre_id` do decorator do onchange do ncm.


![image](https://github.com/OCA/l10n-brazil/assets/634278/17461277-b535-46c0-b077-a3d71fa9a98f)

crsilveira commented:
..... vi que já existe um onchange exclusivo para o fiscal_type  ..........

sim, por isso minha dúvida. Obrigado pela resposta vou ver.

rvalyi commented:
outro detalhe: vc pode fazer um git push com a opção -f (force) depois que faz rebase. Se fizer push com -f na mesma branch, vc não precisa criar outro PR, pode ser no mesmo PR. Alias vou re-abrir caso vc quiser fazer isso que ai é melhor as explicações do porque do commit ficam todas no mesmo PR centralizado...

crsilveira committed:
ação já executada no onchange_fiscal_type

marcelsavegnago commented:
Tem algo errado.. faz um rebase com a branch 14 do remote original

renatonlima reviewed (changes requested):
@crsilveira 


Eu vi que o método _onchange_ncm_id tem o decorador @api.onchange("ncm_id", "fiscal_genre_id") que poderia ser alterado removendo o campo fiscal_genre_id, porque não faz sentido agora, então o método ficaria assim:

```python
@api.onchange("ncm_id")
def _onchange_ncm_id(self):
    for r in self:
        if r.ncm_id:
            r.fiscal_genre_id = self.env["l10n_br_fiscal.product.genre"].search(
                [("code", "=", r.ncm_id.code[0:2])]
            )
```

A ideia original desse método era que quando alterado o fiscal_genre_id para o "00" Serviço, fosse alterado o campo ncm_id para o NCM de serviço 000.00.00, mas como reportado pelo issue, não estava correto.

Eu diria que poderia ser complementado nesse PR era que se for alterado o campo fiscal_genre_id e caso o NCM esteja preenchido com o NCM de outra família, poderia ser limpado o campo para que o usuário preencha o NCM correto

```python
@api.onchange("fiscal_genre_id")
def _onchange_ fiscal_genre_id(self):
    for r in self:
        if r. fiscal_genre_id and r.ncm_id:
            if r.fiscal_genre_id.code != r.ncm_id.code[0:2]:
                r.ncm_id = False
```

Além esse método também poderia ser adicionado uma constraint para fazer essa validação (garantir que ao salvar o produto o ncm_id seja da mesma família do fiscal_genre_id, mas acho que isso poderia ficar para depois e ser pensado melhor, porque acredito que esse tipo de validação tem o potencial de causar alguns problemas em uma base de dados multi-company com uma empresa que não é do Brasil.

marcelsavegnago commented:
@crsilveira se puder fazer um rebase e depois dos ajustes puder fazer um squash dos commits.. seria bom. Thanks.

crsilveira committed:
arrumado onchange ncm e adicionado o fiscal_genre conforme sugestao Renato.

antoniospneto commented:
Substituido por #3469
